### ISSUE TITLE
Refactor ItemFactory into hassle-free universal item/prefab spawn system B: 100

### ISSUE THREAD

**Issue by fomalsd** [unitystation/unitystation#1488]
## Category
TODO

## Details

Refactor ItemFactory into either universal static factory and/or implement custom constructors for scripts, so that all spawned items would just work (have all parameters correct on both server and clientside)
That will help create a convenient in-game item spawner later on (and maybe even allow drawing items with tilemap tools in runtime).

As @chairbender mentioned a while ago:
> I'm noticing an issue with constructing stuff from prefabs. Because we are using an object pool pattern to instantiate things from prefabs, sometimes we are not instantiating our objects in a "fully initialized" state. There's a risk that a developer creating a new object from a prefab will not fully initialize all the required state. For example if creating a bullet, they might fail to set some property of it like the behavior's damage amount. A static factory would avoid this issue (and still allow us to use an object pool) by ensuring that any needed state is provided when the object is grabbed from the pool. Currently we seem to generally be relying on so called "procedural instantiation", which has some downsides compared to a constructor / static factory.

It's a relatively big and delicate task that will require a lot of thorough multiplayer testing.

**Comment by DooblyNoobly** [comment]
Bounty of 100 usd will be awarded to the contributor on a successful merge of a solution to this request

**Comment by chairbender** [comment]
I'm having trouble understanding what this ticket is asking for. 

Can someone provide a specific example of how this might look post-refactoring? Like a code example of how you'd like to be able to use the new system to spawn an item, vs. what we have currently.

Let's say you'd want to spawn a pistol magazine.
Currently to spawn an item you use ItemFactory.SpawnItem(magazinePrefab, somePosition, someParent).

I think the downside we are trying to point it is that this doesn't pass any initial state to the item other than things that are common to all items.

I don't think a static factory would be the way to go as far as being able to define all of the required state of the item to be spawned. You would need a method per item. Instead, I would leave it up to an object that captures the state of the item's class itself.

So instead, maybe you'd want to be able to do something like this:
An interface is defined called ISpawnableItem, which defines a "Spawn" method that accepts a position and parent.
We have a class called PistolMagazine which implements this interface.
PistolMagazine mag = new PistolMagazine(initialAmmo);
ItemFactory.Spawn(mag, position, parent);
Internally in ItemFactory.Spawn it invokes the PistolMagazine's implementation of ISpawnableItem.Spawn, along with whatever else it needs to do.

I of course am not suggesting my example is the perfect way of achieving this ticket, would take some more iteration to get a better system in place.

If someone can help me understand also by suggesting the problems with the current approach that we are trying to fix, and what would be desired by the new system, that would help. I can start such a list:

Current downsides:
- Items can be spawned without specifying all of their needed state.
- It is not clear without digging into the code what state needs to be set on a given item in order for it to function properly.
- Due to the need to expose an item's state publicly in order to initialize it, too much of the item's internal state is exposed and publicly mutable.

Desired features:
- Can only spawn an item in a fully-initialized state, and it is straightforward to figure out how to spawn an item in such a way.
- Item state is not excessively publicly mutable.
- Want to be able to have an item-spawner in game later on which would allow items to be instantiated within the game, with the user being able to specify all of the needed data for the initial state of the item. This does not have to be completed as part of this ticket I would think.

So to summarize, looking for the following info:
1. Give a specific example of how this new system might be used to spawn an item, vs the current approach.
2. Help me flesh out the lists above.

**Comment by chairbender** [comment]
More suggestions from Bod9001:
I think it's mainly of 
Making sure items are initialised properly
Making it easy to recycle them
Easy spawning
Easy making
Reducing the number of prefabs(?)

**Comment by fomalsd** [comment]
As an end user I'd like to see a universal spawn menu with a searchable list of all possible spawnable items (prefabs and uni-items alike) and objects (machines, crates, whatever), so this system should allow spawning anything without any context.
Another thing revamped ItemFactory IMO should be capable of is cloning existing objects with all their metadata.

**Comment by chairbender** [comment]
Taking this on
